internal/pkg/chunker/file/file.go

```diff
@@ -57,88 +56,97 @@ func NewChunker(source Source, setters ...Option) chunker.Chunker {
 }
 
 // Read implements ChunkReader.
+//
+// nolint: gocyclo
 func (c *File) Read(ctx context.Context) <-chan []byte {
 	// Create a buffered channel of length 1.
 	ch := make(chan []byte, 1)
 
 	go func(ch chan []byte) {
 		defer close(ch)
 
+		watcher, err := fsnotify.NewWatcher()
+		if err != nil {
+			log.Printf("failed to watch: %v\n", err)
+			return
+		}
+		// nolint: errcheck
+		defer watcher.Close()
+
+		if err = watcher.Add(filepath.Dir(c.source.Name())); err != nil {
+			log.Printf("failed to watch add: %v\n", err)
+			return
+		}
 		offset, err := c.source.Seek(0, io.SeekStart)
 		if err != nil {
+			log.Printf("failed to seek: %v\n", err)
 			return
 		}
+
 		buf := make([]byte, c.options.Size)
+
 		for {
-			select {
-			case <-ctx.Done():
-				return
-			default:
+			for {
 				n, err := c.source.ReadAt(buf, offset)
-				if err != nil {
-					if err != io.EOF {
-						fmt.Printf("read error: %s\n", err.Error())
-						break
-					}
+				if err != nil && err != io.EOF {
+					log.Printf("read error: %s\n", err.Error())
+					return
 				}
+
 				offset += int64(n)
-				if n != 0 {
+
+				if n > 0 {
 					// Copy the buffer since we will modify it in the next loop.
 					b := make([]byte, n)
 					copy(b, buf[:n])
-					ch <- b
 
-					// Clear the buffer.
-					for i := 0; i < n; i++ {
-						buf[i] = 0
+				DELIVER:
+					select {
+					case <-ctx.Done():
+						return
+					case event := <-watcher.Events:
+						// drain events while waiting for the buffer to be delivered
+						// otherwise inotify() queue might overflow
+						if event.Name == c.source.Name() && event.Op == fsnotify.Write {
+							// clear EOF condition (if there was one) to make sure
+							// we read more data
+							err = nil
+						}
+						goto DELIVER
+					case ch <- b:
 					}
-				} else if err := watch(c.source.Name()); err != nil {
-					log.Printf("failed to watch: %v\n", err)
-					return
 				}
-			}
-		}
-	}(ch)
 
-	return ch
-}
-
-func watch(path string) error {
-	watcher, err := fsnotify.NewWatcher()
-	if err != nil {
-		return err
-	}
-	// nolint: errcheck
-	defer watcher.Close()
-
-	errChan := make(chan error)
+				if err == io.EOF || n == 0 {
+					break
+				}
+			}
 
-	go func() {
-		for {
+		WATCH:
 			select {
+			case <-ctx.Done():
+				return
 			case event := <-watcher.Events:
+				if event.Name != c.source.Name() {
+					// ignore events for other files
+					goto WATCH
```

Comment:
why implement `goto` instead of `continue`?

Reply:
hard to say, these loops/selects were pretty complicated, I thought `goto` is more explicit in terms of control flow, but I can change to `continue`, whatever looks better for you, guys

Reply:
two nested for loops, plus select, I think I created a monster ;)

Reply:
[break](https://golang.org/ref/spec#Break_statements) is a bit ambiguous as it terminates `select` and `for`, and `continue` seems to be a friend of `break`, but it does nothing in `select`

Reply:
I tried to rebuild the code to use `for` instead of `goto`, but looks like it still requires labels, as I have to `break` out of `for` which has nested inside `select`

I haven't got any better idea so far.

Reply:
We can live with it for now and think about it again when we have more time. It works at least now :D